core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -467,7 +467,9 @@ private[spark] class Master(
    * two executors on the same worker).
    */
   def canUse(app: ApplicationInfo, worker: WorkerInfo): Boolean = {
-    worker.memoryFree >= app.desc.memoryPerSlave && !worker.hasExecutor(app)
+    worker.memoryFree >= app.desc.memoryPerSlave &&
+    !worker.hasExecutor(app) &&
+    !app.removedExecutors.exists(_.worker == worker)
```

Comment:
I don't think this is a good idea as is. You can have an executor that died because of transient failures(network blips) and if you blacklist them, they'll never be able to come back. This means that your cluster will monotonically decrease in size even if the once bad executors can in theory be relaunched (unless you add more hosts dynamically). If you want to do this, i'd recommend having a TTL on the blacklist (has been removed since a specific time).